### DISCUSSION TITLE
How to use modified Vim plugins (e.g. Termdebug)

### DISCUSSION THREAD

**Discussion by jremus** [vim/vim#16035] [Q&A]
I would like to use a customized version of the Termdebug plugin, which subsitutes tabs and newlines in balloon. For that purpose I have checked out the Vim source code to `~/git/vim/`, made modifications in a personal branch, and created a symlink `~/.vim/pack/vim/opt/termdebug/plugin/termdebug.vim` to `~/git/vim/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim`. In my `~/.vimrc` I had:
```
packadd termdebug
```
This worked flawlessly for some time until recently when upgrading from Ubuntu 24.04 to 24.10 I now get the following message when starting Vim:
```
[termdebug] Termdebug already loaded.
Press ENTER or type command to continue
```
The fix from issue #15256 to use `packadd! termdebug` does not resolve my issue.

For testing I removed my modified copy of termdebug.vim (by renaming `~/.vim/pack/` to `~/.vim/pack-renamed/`). Using `packadd termdebug` I then indeed run into the issue described in issue #15256. Using `packadd! termdebug` this issue then goes away. But so far I could not come up with how to use my personal modified copy instead without triggering this annoying message.

What is the preferred approach to replace selected Vim plugins with custom modified versions? How do developers use their modified versions?

Thanks and regards,
Jens

**Comment by benknoble** [comment]
I don’t know if this would work for termdebug, but I suspect it might: you could change the package name (say, to my-termdebug).

You could even version-control your forked termdebug sources and use [subtree merges](https://web.mit.edu/git/www/howto/using-merge-subtree.html) ([source](https://github.com/git/git/blob/master/Documentation/howto/using-merge-subtree.txt)) if you wanted to contribute upstream to Vim (I’m not sure there’s an easy way to go the other direction, but there might be).

**Comment by ubaldot** [comment]
Instead of using `packadd! termdebug`, you can place a customized copy of `termdebug.vim` wherever you want and just source it.
That is, you can replace the `packadd! termdebug` from your `.vimrc` and use `source /path/to/your/custom/termdebug/termdebug.vim.` instead.

Nevertheless, if you see such an annoying message it means that termdebug is already loaded, most likely from some other script. 
Or, you see it when you source your .vimrc more than once, and therefore you try to re-load termdebug which is already loaded. 
You can check if termdebug is loaded with `echo g:termdebug_loaded`. If this is true and you try to reload the plugin, then you get that message.  I am not sure if we should silence it though. 
Anyhow, you can get rid off that message in your customized version by commenting or removing the following line (it should be line 52):

` Echoerr('Termdebug already loaded.')`

Regarding how variables are displayed, let me mention this recent PR: https://github.com/vim/vim/pull/15933 . Perhaps the same format could be extended to balloons in the main distribution.

**Comment by ubaldot** [comment]
As @benknoble has mentioned, feel free to send PR if you have some ideas for improving termdebug. :)
